### DISCUSSION TITLE
关于自定义组件的问题

### DISCUSSION THREAD

**Discussion by kuangyuliang** [alibaba/formily#1780] [Q&A 2.x] [answered]
我使用formily 2.0作为流程节点的属性，因为节点类型不同会动态切换schema，在自定义组件里面需要请求后端API，在切换schema的时候自定义组件会unmount，这样我再切换回来的时候又需要重新请求一次数据，请问有什么好的办法解决这个问题么？
例子：https://codesandbox.io/s/dazzling-brown-edj95?file=/App.tsx:425-431

**Comment by 729993031** [comment]
你可以使用一个外部对象缓存你请求的数据

**Reply by kuangyuliang** [reply]
加缓存是可以的，主要是想在切换schema的时候组件不要销毁就好了，因为有些组件会共用。

**Reply by liuweiGL** [reply]
整个表单的状态是维护在 form 实例里面的，组件销毁了没任何影响。

**Comment by liuweiGL** [comment, marked as answer]
需要看你的业务需求来决定请求的时机：
1. 请求跟业务关联的：比如 getDataByXXX 接口中，如果 xxx 参数是组件的 props 或 state，那么这个请求应该放在 useEffect 中。
```tsx
const Test = (props) =>{
    useEffect(()=>{
         getDataByXXX(props.xxx)
    },[props.xxx])
}
```
2. 请求跟业务没有关联：比如 getDicts 就是从后端获取字典数据，那么可以把请求放在组件外部。
```tsx
const prefetcher = getDicts()

const Test = ()=>{
const [dicts, setDicts] = useState()

useEffect(()=>{
    prefetcher.then(setDicts)
},[])
}
```
3. 在 2 中如果不想预请求，可以写个简单的 lazy 加载器

```tsx
const createFetcher = api => {
   let fetcher

   return ()=>{
        if(!fetcher){
           fetcher = api()
        }
       return fetcher
    }
}

const dictFetcher = createFetcher(getDicts)

const Test = ()=>{
    const [dicts, setDicts] = useState()
    
    useEffect(()=>{
        dictFetcher().then(setDicts)
    },[])
}
```

**Reply by kuangyuliang** [reply]
在自定义组件内部请求的数据，切换schema之后组件就销毁了，数据就没了，下次再加载这个schema的时候需要重新加载数据

**Reply by liuweiGL** [reply]
你说的这个问题 2/3 方案不能解决吗？

**Reply by kuangyuliang** [reply]
你可以看看我写的那个例子，请求数据没什么问题，就是我这个数据只需要组件加载的时候请求一次，在切换别的schema再切换回来的时候useState里面的数据也会清空，又会重新执行mount，所以又会重新加载一次数据，除非缓存这次请求的数据，在mount阶段判断是否还需要重新请求。

**Reply by liuweiGL** [reply]
建议你仔细看看我的回答，就是只请求一次啊。

**Reply by kuangyuliang** [reply]
不好意思，没仔细看，谢谢解答！
